### PR TITLE
Adding Sender/Receiver functionality to transport/dummy transport.

### PR DESCRIPTION
**Pull request by err508**
This adds a source/target logic to `Dummy_transport.send_message` to make it more usable.

### CODE REVIEW

File: raiden_libs/transport/transport.py [hunk 1]
```diff
@@ -19,12 +21,12 @@ class Transport(gevent.Greenlet):
...
-        self.message_callbacks = list()
+        self.message_callbacks = defaultdict(list)
 
-    def add_message_callback(self, callback):
-        self.message_callbacks.append(callback)
+    def add_message_callback(self, callback, target_node: str = None):
+        self.message_callbacks[target_node].append(callback)
 
-    def run_message_callbacks(self, data):
+    def run_message_callbacks(self, data, target_node: str = None):
```
Comment (pcppcp) [inline raiden_libs/transport/transport.py]: maybe update this with a simple broadcast option: If `target_node is None`, deliver message to all nodes.

File: raiden_libs/transport/matrix.py [hunk 2]
```diff
@@ -73,7 +73,7 @@ def dispatch(self, room, event):
                 self.run_message_callbacks(event['content']['body'])
                 log.debug("{0}: {1}".format(event['sender'], event['content']['body']))
 
-    def transmit_data(self, message: str, target_node: str = None):
+    def transmit_data(self, message: str, target_node: str = None):  # type: ignore
```
Comment (palango) [inline raiden_libs/transport/matrix.py]: why dis `type: ignore` needed here.

Comment (palango) [reply raiden_libs/transport/matrix.py]: That doesn't make sense. Then you will get errors when you call the `send_message` function on the `MatrixTransport`.

Comment (err508) [reply raiden_libs/transport/matrix.py]: Nope, default in transport is none, so no problem when calling it. The problem is only non-matching signatures for type-checking. One could also add a `source_node = None` to matrix, but that wouldn't make sense, as in matrix logic the source node is known.

Comment (palango) [reply raiden_libs/transport/matrix.py]: This simplified script gives me the following error:
```py

class Transport:
    def send_message(
            self,
            message: str,
            source_node: str = None,
            target_node: str = None
    ):
        self.transmit_data(message, source_node, target_node)

    def transmit_data(self, message: str, source_node: str = None, target_node: str = None):
        raise NotImplementedError

class TestTransport(Transport):
    def transmit_data(self, message: str, target_node: str):
        print('message', message)
        print('target', target_node)


a = TestTransport()
a.send_message('message', 'from', 'to')
```

```
% python3 derive.py                                                                                                                                                                  10:55:48
Traceback (most recent call last):
  File "derive.py", line 21, in <module>
    a.send_message('message', 'from', 'to')
  File "derive.py", line 9, in send_message
    self.transmit_data(message, source_node, target_node)
TypeError: transmit_data() takes 3 positional arguments but 4 were given
```

So from my understanding this doesn't work.

Comment (err508) [reply raiden_libs/transport/matrix.py]: Okay I see, I'll push the changes to DummyTransport than instead so we don't have to change anything in Matrix.

File: raiden_libs/test/mocks/dummy_transport.py [hunk 3]
```diff
@@ -1,26 +1,82 @@
...
         self.sent_messages = defaultdict(list)
 
     def _run(self):
         self.is_running.wait()
 
-    def receive_fake_data(self, data: str):
+    def add_message_callback(self, callback, **kwargs):
+        self.message_callbacks[kwargs["target_node"]].append(callback)
+
+    def run_message_callbacks(self, data, **kwargs):
```
Comment (palango) [inline raiden_libs/test/mocks/dummy_transport.py]: Why does this need to be in the DummyTransport?

**Comment by palango** [comment]
In my opinion it just became worse. Not there's no structure or information for knowing what kind of parameters a function expects. Every `kwargs['bla']` will fail with an `KeyError` when an expected parameter is not supplied.

**Comment by err508** [comment]
Back to version from today, 12:30

File: raiden_libs/test/mocks/dummy_transport.py [hunk 4]
```diff
@@ -1,26 +1,87 @@
...
+            log.error('Error when deserializing message: %s', str(ex))
+            return
+
+        for callback in self.message_callbacks[target_node]:
+            callback(deserialized_msg)
+
+    def send_message(  # type: ignore
+            self,
+            message: Union[str, Message],
+            source_node: str = None,
```
Comment (palango) [inline raiden_libs/test/mocks/dummy_transport.py]: `target_node` should be the second parameter, as in `Transport`

File: raiden_libs/test/mocks/dummy_transport.py [hunk 5]
```diff
@@ -1,26 +1,87 @@
...
-    def transmit_data(self, data: str, target_node: str = None):
+    def transmit_data(  # type: ignore
+            self,
+            data: str,
+            source_node: str = None,
+            target_node: str = None
+    ):
         """Implements `transmit_data` method of the `Transport` class. """
-        self.sent_messages[target_node].append(data)
+        self.sent_messages[source_node].append(data)
```
Comment (palango) [inline raiden_libs/test/mocks/dummy_transport.py]: this still shouldn't trigger callbacks

**Pull request closed without merging**